contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,120 @@
     uint256 public constant MAXIMUM_DELAY = 30 days;
 
+    mapping(bytes32 => TrnasactionStatus) public queuedTransactions;
+
     address public admin;
-    address public pendingAdmin;
     uint256 public delay;
 
-    mapping(bytes32 => bool) public queuedTransactions;
+    event QueuedTransactionsBatch(bytes32 indexed txDataHash, uint256 numberOfTransactions, uint256 eta);
+    event CancelledTransactionsBatch(bytes32 indexed txDataHash);
+    event ApprovedTransactionsBatch(bytes32 indexed txDataHash);
+    event ExecutedTransaction(bytes32 indexed txDataHash, address indexed target, uint256 value, string signature, bytes data, uint256 eta);
+    event NewAdminSet(address indexed newAdmin);
 
-    constructor(address admin_, uint256 delay_) {
-        require(admin_ != address(0), "Admin address cannot be 0");
-        require(delay_ >= MINIMUM_DELAY, "Timelock::constructor: Delay must exceed minimum delay.");
-        require(delay_ <= MAXIMUM_DELAY, "Timelock::setDelay: Delay must not exceed maximum delay.");
+    constructor(address _admin, uint256 _delay) {
+        require(_admin != address(0), "Admin address cannot be 0");
+        require(_delay >= MINIMUM_DELAY, "Timelock: Delay must exceed minimum delay.");
+        require(_delay <= MAXIMUM_DELAY, "Timelock: Delay must not exceed maximum delay.");
 
-        admin = admin_;
-        delay = delay_;
+        admin = _admin;
+        delay = _delay;
     }
 
-    function setDelay(uint256 delay_) external {
-        require(msg.sender == address(this), "Timelock::setDelay: Call must come from Timelock.");
-        require(delay_ >= MINIMUM_DELAY, "Timelock::setDelay: Delay must exceed minimum delay.");
-        require(delay_ <= MAXIMUM_DELAY, "Timelock::setDelay: Delay must not exceed maximum delay.");
-        delay = delay_;
+    function setAdmin(address _admin) onlyOwner external {
+        require(_admin != address(0), "Admin address cannot be 0");
+        
+        admin = _admin;
 
-        emit NewDelay(delay);
+        emit NewAdminSet(admin);
     }
 
-    function acceptAdmin() external {
-        require(msg.sender == pendingAdmin, "Timelock::acceptAdmin: Call must come from pendingAdmin.");
-        admin = msg.sender;
-        pendingAdmin = address(0);
-
-        emit NewAdmin(admin);
-    }
+    function queueTransactionsBatch(Transaction[] memory transactions, uint256 eta) onlyAdmin external returns (bytes32) {
+        require(
+            msg.sender == admin,
+            "Timelock: Call must come from admin."
+        );
 
-    function setPendingAdmin(address pendingAdmin_) external {
-        require(msg.sender == address(this), "Timelock::setPendingAdmin: Call must come from Timelock.");
-        pendingAdmin = pendingAdmin_;
+        //todo ag hmm... do we really want it this way?
+        require(eta >= block.timestamp + delay, "Timelock: Estimated execution block must satisfy delay."); //todo add test!
```

Comment:
I think in general the idea is to have multiple timelocks. One with a minimum of lets say 48 hours - used for governance decisions when needed (this will give investors the confidence these kinds of decisions will take time to be executed).
Another timelock would be for immediate actions.

I saw that Sovryn did something similar..

WDYT?

Reply:
TimeLock contract here is the owner of all of the other contracts. I'm not sure how to implement it in our case. We would need to add roles to the contracts which would have different privileges and delays

Reply:
@kazazor if we will have those two  timelocks it basically means we have no timelock since we are able to execute any upgrade immediately (its up to us to decide which timelock to use for each action we are running). I guess we must set some HARD minimum timelock so it will be considered safe

Reply:
That's a problem with emergency tasks.. like the update oracles

Maybe we could have a map of minimum delay?

Reply:
>Maybe we could have a map of minimum delay?

@kazazor It would have the same effect that @tkeidar mentioned. There is no timelock if you can bypass it

Reply:
Maybe we should check what other protocols did.. There are protocols with emergency functions, I'm sure it doesn't wait for a minimum timelock if needed to pause for example..

Maybe only certain functions are excluded from the timelock..